File: agni-core/src/main/kotlin/dev/agni/core/protocol/Command.kt
package dev.agni.core.protocol

sealed class Command {
    object Ping : Command()
    object Healthcheck : Command()
    data class Get(val key: String) : Command()
    data class Set(val key: String, val value: ByteArray) : Command()
    data class Unknown(val message: String) : Command()

    companion object {
        // Mirrors the Rust parser: split on the first two spaces only, so a
        // SET value can contain spaces of its own.
        fun fromBytes(bytes: ByteArray): Command {
            val input = String(bytes, Charsets.UTF_8)
            val parts = input.split(' ', limit = 3)
            val command = parts.getOrElse(0) { "" }.uppercase()

            return when (command) {
                "PING" -> Ping
                "HEALTHCHECK" -> Healthcheck
                "GET" -> {
                    val key = parts.getOrNull(1)
                    if (key != null) Get(key.trim()) else Unknown("GET requires a key")
                }
                "SET" -> {
                    val key = parts.getOrNull(1)
                    val value = parts.getOrNull(2)
                    if (key != null && value != null) {
                        Set(key.trim(), value.trim().toByteArray(Charsets.UTF_8))
                    } else {
                        Unknown("SET requires a key and value")
                    }
                }
                else -> Unknown(command)
            }
        }
    }
}

File: go-proto/internal/protocol/command.go
package protocol

import "strings"

type CommandKind int

const (
	Ping CommandKind = iota
	Healthcheck
	Get
	Set
	Unknown
)

// Command mirrors agni-core's sealed Command class as a tagged struct:
// Key/Value/Message are only populated for the variants that use them.
type Command struct {
	Kind    CommandKind
	Key     string
	Value   []byte
	Message string
}

// ParseCommand mirrors the Rust/Kotlin parser: split on the first two
// spaces only, so a SET value can contain spaces of its own.
func ParseCommand(data []byte) Command {
	parts := strings.SplitN(string(data), " ", 3)
	name := strings.ToUpper(parts[0])

	switch name {
	case "PING":
		return Command{Kind: Ping}
	case "HEALTHCHECK":
		return Command{Kind: Healthcheck}
	case "GET":
		if len(parts) < 2 {
			return Command{Kind: Unknown, Message: "GET requires a key"}
		}
		return Command{Kind: Get, Key: strings.TrimSpace(parts[1])}
	case "SET":
		if len(parts) < 3 {
			return Command{Kind: Unknown, Message: "SET requires a key and value"}
		}
		return Command{
			Kind:  Set,
			Key:   strings.TrimSpace(parts[1]),
			Value: []byte(strings.TrimSpace(parts[2])),
		}
	default:
		return Command{Kind: Unknown, Message: name}
	}
}
